core/src/main/java/org/elasticsearch/action/support/replication/ReplicatedOperation.java

```diff
@@ -0,0 +1,334 @@
+        // can be null if the index is deleted / closed on us..
+        // nocommit: I'm not sure this is the right solution here. Deletion is easy (it doesn't matter), closing an index
+        // during replication may result in loss of acked writes. Solving this will required a bigger effort around closed
+        // indices.
+        final IndexShardRoutingTable shardRoutingTable = state.getRoutingTable().shardRoutingTableOrNull(shardId);
+        List<ShardRouting> shards = shardRoutingTable == null ? Collections.emptyList() : shardRoutingTable.shards();
+        return shards;
+    }
+
+    private void decPendingAndFinishIfNeeded() {
+        if (pendingShards.decrementAndGet() <= 0) {
+            finish();
+        }
+    }
+
+    private void finish() {
+        if (finished.compareAndSet(false, true)) {
+            final ReplicationResponse.ShardInfo.Failure[] failuresArray;
+            if (shardReplicaFailures.isEmpty()) {
+                failuresArray = ReplicationResponse.EMPTY;
+            } else {
+                failuresArray = new ReplicationResponse.ShardInfo.Failure[shardReplicaFailures.size()];
+                shardReplicaFailures.toArray(failuresArray);
+            }
+            finalResponse.setShardInfo(new ReplicationResponse.ShardInfo(
+                            totalShards.get(),
+                            successfulShards.get(),
+                            failuresArray
+                    )
+            );
+            finalResponseListener.onResponse(finalResponse);
+        }
+    }
+
+    private void finishAsFailed(Throwable throwable) {
+        if (finished.compareAndSet(false, true)) {
+            finalResponseListener.onFailure(throwable);
+        }
+    }
+
+
+    /**
+     * Should an exception be ignored when the operation is performed on the replica.
+     */
+    public static boolean ignoreReplicaException(Throwable e) {
+        if (TransportActions.isShardNotAvailableException(e)) {
+            return true;
+        }
+        // on version conflict or document missing, it means
+        // that a new change has crept into the replica, and it's fine
+        if (isConflictException(e)) {
+            return true;
+        }
+        return false;
+    }
+
+    public static boolean isConflictException(Throwable e) {
+        Throwable cause = ExceptionsHelper.unwrapCause(e);
+        // on version conflict or document missing, it means
+        // that a new change has crept into the replica, and it's fine
+        if (cause instanceof VersionConflictEngineException) {
+            return true;
+        }
+        return false;
+    }
+
+
+    interface Primary<Request extends ReplicationRequest<Request>, ReplicaRequest extends ReplicationRequest<ReplicaRequest>,
+            Response extends ReplicationResponse> {
+
+        /** returns the current primary term */
+        long term();
+
+        ShardRouting routingEntry();
```

Comment:
It feels weird to me to pass the routing entry as a parameter. As discussed we could maybe move all of this to use allocationIds. However that will require more changes (for example to shard status action, to fail shards based on allocation Ids) and will also make logging harder to read (the routing show more info about which shard this is). Alos, the replicas are based keyed by routing because it gives a directed access to the relevant node. All in all I rather not do it now.